### PR TITLE
Treat TS-only keywords as identifiers in jsdoc comments

### LINKED ISSUE #8222: javascript doc comments don't work for a parameter named `type`
_From @epicgirl1998 on April 20, 2016 13:44_

- VSCode Version: 1.0.0
- OS Version: Windows 8.1

Steps to Reproduce:

1. create a .js file containing the snippet below
2. hover over the function's name

```javascript
/**
 * @param {string} type
 */
function test(type) {
	return true;
}
```

the type hint for `type` should be `string`, but instead it is `any`

other parameter names do not suffer from the issue

i made a screenshot for your viewing pleasure:

type

the dev tools console is empty and i don't have any extensions installed

_Copied from original issue: Microsoft/vscode#5549_

### PR DESCRIPTION
**Pull request by zhengbli**
Fixes #8222 
The list on TS-only keywords is created with references to 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar
and 
http://www.w3schools.com/js/js_reserved.asp

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -1667,6 +1667,27 @@ namespace ts {
         return SyntaxKind.FirstKeyword <= token && token <= SyntaxKind.LastKeyword;
     }
 
+    // Some keywords are TypeScript only, so they should not be parsed as considered as
+    // keywords in JavaScript
+    export function isJSKeyword(token: SyntaxKind): boolean {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: You only use this for `isTSOnlyKeyword`, so why not just write

```ts
function isSolelyTypeScriptKeyword(token: SyntaxKind) {
    switch (token) {
        case SyntaxKind.
        case SyntaxKind.
    }
}

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -1667,6 +1667,27 @@ namespace ts {
...
+    // keywords in JavaScript
+    export function isJSKeyword(token: SyntaxKind): boolean {
+        switch (token) {
+            case SyntaxKind.TypeKeyword:
+            case SyntaxKind.AsKeyword:
+            case SyntaxKind.AnyKeyword:
+            case SyntaxKind.DeclareKeyword:
+            case SyntaxKind.IsKeyword:
+            case SyntaxKind.ReadonlyKeyword:
+            case SyntaxKind.FromKeyword:
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: `from` is a contextual ES6 keyword

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -1667,6 +1667,27 @@ namespace ts {
         return SyntaxKind.FirstKeyword <= token && token <= SyntaxKind.LastKeyword;
     }
 
+    // Some keywords are TypeScript only, so they should not be parsed as considered as
+    // keywords in JavaScript
+    export function isJSKeyword(token: SyntaxKind): boolean {
+        switch (token) {
+            case SyntaxKind.TypeKeyword:
+            case SyntaxKind.AsKeyword:
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: `as` is a contextual ES6 keyword for imports

File: src/compiler/utilities.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: If you keep it around, use `isSolelyJavaScriptKeyword`.

Additionally, you might want to note that these keywords are language version dependent.

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -6259,7 +6259,7 @@ namespace ts {
                 }
 
                 function parseJSDocIdentifier(): Identifier {
-                    if (token !== SyntaxKind.Identifier) {
+                    if (token !== SyntaxKind.Identifier && !isTSOnlyKeyword(token)) {
```
Comment (mhegazy) [inline src/compiler/parser.ts]: what is wrong with `tokenIsIdentifierOrKeyword(token)`? do JS keywords appear in the grammar in a valid position?

Comment (zhengbli) [inline src/compiler/parser.ts]: We don't want to allow *all* keywords though. If the keyword is a JS keyword, it should be a syntax error.

File: src/compiler/utilities.ts (hunk 1 above)
Comment (zhengbli) [inline src/compiler/utilities.ts]: I was thinking maybe there will be TSX/JSX-only keywords in the future, then `isSolelyTypeScriptKeyword` wouldn't return the same thing as `isJSKeyword`. Though that might not likely to happen

File: src/compiler/parser.ts (hunk 4 above)
Comment (mhegazy) [inline src/compiler/parser.ts]: well.. not sure i see much value..

`/** param {myNonExistingTypeName} x **/` is not an error. so why is it different from 
`/** param {import} x **/`

Comment (zhengbli) [inline src/compiler/parser.ts]: `parseJSDocIdentifier` doesn't work inside a type expression. The difference cases should be:

```js
/** param {myNonExistingTypeName} x **/ // ok
/** param {myNonExistingTypeName} import **/ // not ok
/** param {myNonExistingTypeName} type **/ // should be ok, but not now
```

Comment (mhegazy) [inline src/compiler/parser.ts]: thanks for the clarification, but the question still, does it make much difference?

**Comment by zhengbli** [comment]
Discussed with Mohamed offline, we didn't find a case where allow keywords in jsdoc comments would result in errors. So it is simpler to allow both identifiers and keywords in jsdoc instead of maintaining a list of TS-only keywords.

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**